Restoring components architecture document (#2308)
ros2/rclcpp: rclcpp_components/doc/architecture.rst

Architecture of ``rclcpp_components``
=====================================

``rclcpp_components`` provides tools and defines conventions for building ``rclcpp`` nodes into dynamically loadable components.
Building composable nodes makes it easier to compose them with other composable nodes at runtime.
For example, it allows you to build your nodes one time but then at runtime decide whether or not to run them in their own process or combine them into a single process.

It also provides some generic tools for working with composable nodes, including a composable node container, which can load and execute multiple nodes in a variety of scenarios.

Terminology
-----------

- Composable Node:
  - An ``rclcpp`` based node that is setup so that it can be composed into a single process with other composable nodes at either runtime or programming time.
- Node Component:
  - Same thing as Composable Node, can be used interchangeably.
- Composable Node Container:

>   - A program that can load and host composable nodes, it is often general purpose and has not a priori knowledge about the nodes it is instantiating and executing.

Reviewer: ```suggestion
  - A program that can load and host composable nodes, which is often general purpose and has no a priori knowledge about the nodes it is instantiating and executing.
```

- pluginlib:
  - ROS package that provides mechanisms to create and register plugins based on user defined interface classes, and then dynamically lookup and load plugins at runtime.
- Stand-alone Node Executable:

>   - Convenience executable that can be used to run a composable node as a single process or have it loaded dynamically in a remote composable node container.

Reviewer: I don't really understand the latter part of this sentence (I'm not sure what a "remote composable node container" is).  Could you maybe explain a little further or reword it?

Theory of Operation
-------------------

> The lowest level series of steps require to make a composable node are as follows:

Reviewer: ```suggestion
The series of steps require to make a composable node are as follows:
```

(at least, I'm not sure what "lowest level" is trying to convey here)

- Create a shared library in which one or more of your composable nodes are compiled.
  - The nodes should also be registered as a ``pluginlib``-style plugin.
  - The library should be created with CMake's ``add_library``

>   - The library should either explicitly be set to ``SHARED``.

Reviewer: ```suggestion
  - The library should either explicitly be set to ``SHARED``, or be using ``ament_cmake_ros``.
```

- Create or add to a ``plugindescription.xml`` file and register it so that ``pluginlib`` can find the node.
- Create a stand-alone executable for each node.

>   - Which can either be used to run the node in its own process or to load it into an existing process and serving as a proxy while it is in the remote process.

Reviewer: Is the latter part of this sentence true?  Do we keep around a proxy process for composable nodes?

Author: We do not, removing.

Author: Actually, many things in this section aren't relevant, because we aren't using `pluginlib` registration like this anymore.  Let me rework it.

Generally they can be applied in the following way:

- Write files containing nodes, e.g. ``src/a.cpp``, ``src/b.cpp``, etc., and in each use the ``RCLCPP_COMPONENTS_REGISTER_NODE()`` preprocessor macro in the source code.
- In CMake, create the shared library, e.g. ``add_library(my_library SHARED src/a.cpp src/b.cpp)``.
  - Do any extra CMake steps for the library, e.g. ``target_link_libraries`` or ``target_include_directories``.
- Describe each node in the library with the CMake using ``rclcpp_components_describe_node_component``.
  - e.g. ``rclcpp_components_describe_node_component(my_library NAME MyNodeA TYPE my_cpp_namespace::MyNodeA DESCRIPTION "...")``
- Register the library with ``pluginlib`` by calling ``rclcpp_components_register_node_component_library``.
  - e.g. ``rclcpp_components_register_node_component_library(my_library GENERATE_PLUGIN_DESCRIPTION)``
- Optionally, create a stand-alone executable for each node using ``rclcpp_components_create_executable_for_node``.
  - e.g. ``rclcpp_components_create_executable_for_node(my_node_a_target NODE_COMPONENT_NAME MyNodeA)``

>   - An assumption is that package name is equal to ``PROJECT_NAME``, but can be overridden with the ``PACKAGE_NAME <package name>`` option.

Reviewer: I think this set of steps is from an older version.  My understanding is that nowadays, we only need to call `rclcpp_components_register_nodes`, though I could be wrong.

The contract for the class passed to ``RCLCPP_COMPONENTS_REGISTER_NODE`` is that it:

- has a constructor which takes a single argument which is an instance of ``rclcpp::NodeOptions``
- has a method of the signature ``rclcpp::node_interfaces::NodeBaseInterface::SharedPtr get_node_base_interface()``

Note that it does not *need* to inherit from ``rclcpp::Node``, but that is the easiest way to do it and to get the ``get_node_base_interface()`` method for "free".

The RCLCPP_COMPONENTS_REGISTER_NODE Macro
+++++++++++++++++++++++++++++++++++++++++

One possible (psuedo-code) implementation for the ``RCLCPP_COMPONENTS_REGISTER_NODE`` macro might be:

.. code-block:: cpp

    #define RCLCPP_COMPONENTS_REGISTER_NODE(NodeClass) \
      CLASS_LOADER_REGISTER_CLASS(NodeFactoryTemplate<NodeClass>, rclcpp_components::NodeFactory)

Where the base class ``rclcpp_components::NodeFactory`` is the actual plugin interface and the ``rclcpp_components::NodeFactoryTemplate`` class is a templated "adapter" class which will provide the ``rclcpp_components::NodeFactory`` interface given a sub-class of ``rclcpp::Node``.

The NodeFactory and Related Interfaces
++++++++++++++++++++++++++++++++++++++

> The ``rclcpp_components::NodeFactory`` interface is dual-purposed, first it works around a limitation of instantiating classes via a general purpose class loading interface which is that you cannot call constructors on derived classes because you only know the base class and are unaware of any different signatures for constructors in the derived class.

Reviewer: ```suggestion
The ``rclcpp_components::NodeFactory`` interface is dual-purposed.
First, it works around a limitation of instantiating classes via a general purpose class loading interface which is that you cannot call constructors on derived classes because you only know the base class and are unaware of any different signatures for constructors in the derived class.
```

      }

    private:
      std::shared_ptr<void> node_instance_;
      NodeBaseInterfaceGetter node_base_interface_getter_;
    };

This interface only requires that you can give access to the ``rclcpp::node_interfaces::NodeBaseInterface`` pointer, as that's the only thing required to compose (and then execute) composable nodes.
The ability to get access to the original Node instance shared pointer is just for debugging and special cases.
Most of the time the program loading the instance of the node will not be aware of the original type of the node instance, whether it be a ``rclcpp::Node`` or something like ``my_ns::MyNode`` which inherits from ``rclcpp::Node`` or even something else completely.

Finally, the ``rclcpp_components::NodeFactoryTemplate<NodeT>`` convenience class will create an implementation of ``NodeFactory`` based on the ``NodeT`` type.
This template class will work so long as the class given for ``NodeT`` provides the ``get_node_base_interface()`` method that exists on the ``rclcpp::Node`` class.

Building the Shared Library which Contains Composable Nodes
-----------------------------------------------------------

> Typically this is done by just calling the standard ``add_library`` macro in CMake, but any thing that results in generating a CMake target which represents a shared library is sufficient.

Reviewer: ```suggestion
Typically this is done by just calling the standard ``add_library`` macro in CMake, but anything that results in generating a CMake target which represents a shared library is sufficient.
```

Once you've created the target you may manipulate the target with standard CMake macros like ``target_include_directories`` and ``target_link_libraries``.

> Creating the Plugin Description File

Reviewer: Just taking https://github.com/ros2/demos/tree/rolling/composition as an example, I don't see any pluginlib XML description file, either in the sources nor in the `install` directory.  But maybe I'm missing something.

The containers could receive the information about which nodes to load and execute in various ways, but only two are discussed here: a configuration file or via a ROS interface.

In both cases, the information related to loading and executing a node consists the following information (some required and some optional):

- package name
  - name of the package in which the node is located
- node plugin name
  - the name for the plugin that was given in the ``rclcpp_components_describe_node_component()`` or the plugin description XML
- node name
  - overrides the node name hard coded into source code (i.e. like ``__node:=<new name>`` on the command line)
- node namespace
  - overrides the node namespace hard coded into the source code (i.e. like ``__ns:=<new ns>`` on the command line)
- remappings
  - remappings that apply to this node for things like topics and services
- initial parameter values
  - initial values for required parameters or to be used instead of the defaults for optional parameters

Essentially, where to find the node (package name plus node name) and anything that can be influenced with command line arguments to nodes normally.

> Compose Nodes with a Configuration File

Reviewer: Is this currently implemented in the container?  It doesn't seem like it to me, but I could be wrong.
